README.md
# ![/>](./dev/favicon.png) integer_net Magento 2 Module Template

This is a GitHub template.

**Create your module now with the "use this template" button:**

[![Use this template](dev/use-this-template.png)](https://github.com/integer-net/magento2-module-template/generate)

After that, clone it on your machine and run `/dev/init` for automatic configuration. For more information, see [dev/README.md](dev/README.md)

<div align="center"><img src="https://www.integer-net.de/wp-content/uploads/2012/11/firmenprofil.jpg" alt="Waschbär Approved Module" /></div>

---

<!-- TEMPLATE -->

# :module-namespace_:module-name Magento Module
<div align="center">

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-travis]][link-travis]
![Supported Magento Versions][ico-compatibility]

[![Coverage Status][ico-scrutinizer]][link-scrutinizer]
[![Quality Score][ico-code-quality]][link-code-quality]
[![Maintainability][ico-maintainability]][link-maintainability]
</div>

---

:description

## Installation

1. Install it into your Magento 2 project with composer:
    ```
    composer require :vendor/:package
    ```

2. Enable module
    ```
    bin/magento setup:upgrade
    ```

## Configuration

## Usage

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Testing

### Unit Tests

```
vendor/bin/phpunit tests/unit
```

### Magento Integration Tests

0. Configure test database in `dev/tests/integration/etc/install-config-mysql.php`. [Read more in the Magento docs.](https://devdocs.magento.com/guides/v2.4/test/integration/integration_test_execution.html) 

1. Copy `tests/integration/phpunit.xml.dist` from the package to `dev/tests/integration/phpunit.xml` in your Magento installation.

2. In that directory, run
    ``` bash
    ../../../vendor/bin/phpunit
    ```


## Security

If you discover any security related issues, please email :author-email instead of using the issue tracker.

## Credits

- [:author-name][link-author]
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE) for more information.

[ico-version]: https://img.shields.io/packagist/v/:vendor/:package.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/:vendor/:package/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/:vendor/:package?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/:vendor/:package.svg?style=flat-square
[ico-maintainability]: https://img.shields.io/codeclimate/maintainability/:vendor/:package?style=flat-square
[ico-compatibility]: https://img.shields.io/badge/magento-:version-badge-brightgreen.svg?logo=magento&longCache=true&style=flat-square

[link-packagist]: https://packagist.org/packages/:vendor/:package
[link-travis]: https://travis-ci.org/:vendor/:package
[link-scrutinizer]: https://scrutinizer-ci.com/g/:vendor/:package/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/:vendor/:package
[link-maintainability]: https://codeclimate.com/github/:vendor/:package
[link-author]: https://github.com/:author-github
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
### Changed
### Removed
### Fixed


tests/unit/ExampleTest.php
<?php

namespace :module-namespace\:module-name;

use PHPUnit\Framework\TestCase;

class ExampleTest extends TestCase
{
    /** @test */
    public function true_is_true()
    {
        $this->assertTrue(true);
    }
}


dev/tests/TravisJobTest.php
<?php
declare(strict_types=1);

namespace IntegerNet\ModuleTemplate;

use PHPUnit\Framework\TestCase;

class TravisJobTest extends TestCase
{
    /**
     * @dataProvider yamlExpectation
     */
    public function testYamlOutput(array $magentoVersions, string $expectedYaml)
    {
        $this->assertEquals($expectedYaml, TravisJob::getConfiguration($magentoVersions));
    }

    public function yamlExpectation()
    {
        yield 'Magento 2.3 | 2.4' => [
            ['2.3', '2.4'],
            <<<YAML
jobs:
  include:
    -
      php: "7.1"
      env:
        - "MAGENTO_VERSION=2.3"
        - "TEST_SUITE=integration"

    -
      php: "7.2"
      env:
        - "MAGENTO_VERSION=2.3"
        - "TEST_SUITE=integration"

    -
      php: "7.3"
      env:
        - "MAGENTO_VERSION=2.3"
        - "TEST_SUITE=integration"

    -
      php: "7.3"
      env:
        - "MAGENTO_VERSION=2.4"
        - "TEST_SUITE=integration"

    -
      php: "7.4"
      env:
        - "MAGENTO_VERSION=2.4"
        - "TEST_SUITE=integration"
        - "COVERAGE=true"

    -
      php: "7.4"
      env:
        - "MAGENTO_VERSION=2.4"
        - "TEST_SUITE=unit"
        - "COVERAGE=true"

    -
      php: "7.4"
      env:
        - "MAGENTO_VERSION=2.4-develop"
        - "TEST_SUITE=integration"

  allow_failures:
    -
      php: "7.4"
      env:
        - "MAGENTO_VERSION=2.4-develop"
        - "TEST_SUITE=integration"


YAML
    ,
        ];
        yield 'Magento 2.4' => [
            ['2.4'],
            <<<YAML
jobs:
  include:
    -
      php: "7.3"
      env:
        - "MAGENTO_VERSION=2.4"
        - "TEST_SUITE=integration"

    -
      php: "7.4"
      env:
        - "MAGENTO_VERSION=2.4"
        - "TEST_SUITE=integration"
        - "COVERAGE=true"

    -
      php: "7.4"
      env:
        - "MAGENTO_VERSION=2.4"
        - "TEST_SUITE=unit"
        - "COVERAGE=true"

    -
      php: "7.4"
      env:
        - "MAGENTO_VERSION=2.4-develop"
        - "TEST_SUITE=integration"

  allow_failures:
    -
      php: "7.4"
      env:
        - "MAGENTO_VERSION=2.4-develop"
        - "TEST_SUITE=integration"


YAML,
        ];        yield 'Magento 2.3' => [
            ['2.3'],
            <<<YAML
jobs:
  include:
    -
      php: "7.1"
      env:
        - "MAGENTO_VERSION=2.3"
        - "TEST_SUITE=integration"

    -
      php: "7.2"
      env:
        - "MAGENTO_VERSION=2.3"
        - "TEST_SUITE=integration"

    -
      php: "7.3"
      env:
        - "MAGENTO_VERSION=2.3"
        - "TEST_SUITE=integration"
        - "COVERAGE=true"

    -
      php: "7.3"
      env:
        - "MAGENTO_VERSION=2.3"
        - "TEST_SUITE=unit"
        - "COVERAGE=true"


YAML,
        ];
    }
}

tests/integration/ExampleTest.php
<?php

namespace :module-namespace\:module-name;

use PHPUnit\Framework\TestCase;
use Magento\TestFramework\Helper\Bootstrap;

/**
 * @magentoAppIsolation enabled
 */
class ExampleTest extends TestCase
{
    /** @test */
    public function we_can_use_the_object_manager()
    {
        $this->assertInstanceOf(\Magento\TestFramework\ObjectManager::class, Bootstrap::getObjectManager());
    }
}


dev/src/TravisJob.php
<?php
declare(strict_types=1);

namespace IntegerNet\ModuleTemplate;

use Nette\Neon\Encoder;
use Nette\Neon\Neon;
use function array_map as map;

/**
 * Renders YAML for travis job configuration based on supported Magento versions
 */
class TravisJob
{
    /**
     * Latest Magento version. If this version is supported, jobs for the dev branch is added
     */
    private const LATEST_MAGENTO_VERSION = '2.4';
    /**
     * Job configuration for dev branch
     */
    private const DEV_BRANCH_JOBS = [
        [
            'php' => '7.4',
            'env' => [
                'MAGENTO_VERSION=2.4-develop',
                'TEST_SUITE=integration',
            ],
        ],
    ];

    public static function getConfiguration(array $magentoVersions)
    {
        $versionConstraints = Config::getMagentoVersions();
        $jobs = [];
        foreach ($magentoVersions as $magentoKey => $magentoVersion) {
            $phpVersions = map(
                fn($constraint) => ltrim($constraint, '^~'),
                $versionConstraints[$magentoVersion]['php']
            );
            // Integration tests for each PHP version
            foreach ($phpVersions as $phpKey => $phpVersion) {
                $job = [
                    'php' => $phpVersion,
                    'env' => [
                        'MAGENTO_VERSION=' . $magentoVersion,
                        'TEST_SUITE=integration',
                    ],
                ];
                // Integration test coverage only for latest PHP and Magento version
                if ($magentoKey === array_key_last($magentoVersions) && $phpKey === array_key_last($phpVersions)) {
                    $job['env'][] = 'COVERAGE=true';
                }
                $jobs['include'][] = $job;
            }
            // Add unit tests only for latest PHP and Magento version
            if ($magentoKey === array_key_last($magentoVersions)) {
                $jobs['include'][] = [
                    'php' => $phpVersion,
                    'env' => [
                        'MAGENTO_VERSION=' . $magentoVersion,
                        'TEST_SUITE=unit',
                        'COVERAGE=true',
                    ],
                ];
            }
        }
        // Add jobs for dev branch if latest Magento version is supported
        if (in_array(self::LATEST_MAGENTO_VERSION, $magentoVersions, true)) {
            foreach (self::DEV_BRANCH_JOBS as $job) {
                $jobs['include'][] = $job;
                $jobs['allow_failures'][] = $job;
            }
        }
        return str_replace("\t", '  ', Neon::encode(['jobs' => $jobs], Encoder::BLOCK));
    }
}

dev/src/Initialize.php
<?php

namespace IntegerNet\ModuleTemplate;

use Minicli\App;
use Minicli\Command\CommandCall;
use Minicli\Input;
use Minicli\Output\OutputHandler;
use Nette\Neon\Neon;

/**
 * This invokable class is executed from the init script
 */
class Initialize
{
    /**
     * @var Input
     */
    private $prompt;
    /**
     * @var OutputHandler
     */
    private $printer;
    /**
     * @var App
     */
    private App $app;

    public function __construct(App $app)
    {
        $this->app = $app;
        $this->prompt = new Input('>');
        $this->printer = $app->getPrinter();
    }

    public function __invoke(CommandCall $input)
    {
        $this->configureFromFlags($input);

        if ($input->hasFlag('--help') || $input->hasFlag('-h') || $input->hasParam('help')) {
            $this->showHelp();
            return;
        }

        try {
            $rootDir = $this->getRootDir($input);
            $this->printer->display("Module directory: $rootDir");

            $this->welcome();

            $values = Config::getDefaultVariables();
            do {
                $values = $this->askValues($values);
                $this->previewValues($values);
            } while (!$this->confirm('Does that look right?'));
            $values += $this->askMagentoCompatibility();

            $this->replaceValues($rootDir, $values);
            $this->removeReadmeSection($rootDir);
            $this->success($values);
        } catch (\Exception $e) {
            $this->printer->error($e->getMessage());
        }
    }

    private function showHelp()
    {
        $README = 'file://' . dirname(__DIR__) . '/README.md';
        $this->printer->info('The init script replaces all placeholders in the template for you interactively');
        $this->printer->info("You can find an explanation of the placeholders in dev/README.md if needed.");
        $this->printer->display("\t{$README}");
        $this->printer->info('Options:');
        $this->printer->info(
            implode(
                "\n",
                [
                    "\t--unicorn\tUse a more colorful theme",
                    "\t--no-ansi\tDo not use colors",
                    "\t--help | -h\tShow this help",
                ]
            )
        );
    }

    private function configureFromFlags(CommandCall $input): void
    {
        if ($input->hasFlag('--unicorn')) {
            $this->app->setTheme('\Unicorn');
            // Registering a theme initializes a new printer instance
            $this->printer = $this->app->getPrinter();
        }
        if ($input->hasFlag('--no-ansi')) {
            $this->printer->clearFilters();
        }
    }

    private function getRootDir(CommandCall $input): string
    {
        $rootDir = ($input->args[2] ?? '');
        if (strpos($rootDir, '/') !== 0) {
            $rootDir = getcwd() . '/' . $rootDir;
        }
        if (realpath($rootDir) === false) {
            throw new \RuntimeException('Path not found: ' . $rootDir);
        }
        return realpath($rootDir);
    }

    private function welcome()
    {
        $helpUrl = 'https://devdocs.magento.com/guides/v2.4/install-gde/prereq/connect-auth.html';
        $this->printer->info(
            implode(
                "\n",
                [
                    "┌──────────────────────────────────────────────────┐",
                    "│ 🎆 Your open source module will be ready soon! 🎆│",
                    "└──────────────────────────────────────────────────┘",
                    '🔑 First you need a key pair for repo.magento.com. To learn how to obtain keys,',
                    '   see: ' . $helpUrl,
                    '',
                    '❓ Why? The credentials will be stored in auth.json so that services like',
                    '   Scrutinizer can install the required Magento modules with composer.',
                    '',
                    '✏ But you can also leave them empty for now and change auth.json manually later.',
                ]
            )
        );
        $this->printer->error(
            implode(
                "\n",
                [
                    "⚠ The keys will be visible, so do not use an account used for commerce license",
                    "  or for marketplace extensions!"
                ]
            )
        );
    }

    private function askValues(array $defaultValues): array
    {
        $values = [];
        foreach ($defaultValues as $key => $default) {
            $this->printer->out(str_replace('-', ' ', ucfirst(substr($key, 1))) . " [$default] ");
            $values[$key] = $this->prompt->read() ?: $default;
        }
        return $values;
    }

    private function previewValues(array $values): void
    {
        $this->printer->info('🔄 The following values will be replaced in package files:');
        $this->printer->printTable(
            array_merge(
                [['placeholder', 'value']],
                array_map(fn($key, $value) => [$key, $value], array_keys($values), $values)
            )
        );
    }

    private function confirm(string $question): bool
    {
        do {
            $this->printer->out($question . ' [Y/N] ', 'alt');
            $response = $this->prompt->read();
        } while (trim($response) === '');
        return !(stripos($response, 'Y') !== 0);
    }

    private function askMagentoCompatibility(): array
    {
        $this->printer->info(
            implode(
                "\n",
                [
                    '📦 Please specify which Magento versions you intend to support.',
                    '   This will generate:',
                    '   - Version constraints in composer.json',
                    '   - Jobs in .travis.yml',
                ]
            )
        );
        $magentoVersions = [];
        $magentoVersionConstraints = [];
        while (empty($magentoVersions)) {
            foreach (Config::getMagentoVersions() as $version => $constraints) {
                if ($this->confirm("Compatible with Magento {$version}?")) {
                    $magentoVersions[] = $version;
                    $magentoVersionConstraints[] = $constraints;
                }
            }
            if (empty($magentoVersions)) {
                $this->printer->error('Please select at least one Magento version');
            }
        }
        $mergedConstraints = array_map('array_unique', array_merge_recursive(...$magentoVersionConstraints));
        return [
            ':php-constraint'       => implode('||', $mergedConstraints['php']),
            ':framework-constraint' => implode('||', $mergedConstraints['magento-framework']),
            ':version-badge'        => implode('%20|%20', $magentoVersions),
            ':travis-jobs'          => TravisJob::getConfiguration($magentoVersions),
        ];
    }

    private function replaceValues(string $rootDir, array $values): void
    {
        $this->printer->info('Replacing values in files:', true);
        foreach (Config::getFilesToUpdate() as $file) {
            $file = $rootDir . '/' . $file;
            $content = file_get_contents($file);
            file_put_contents($file, strtr($content, $values));
            $this->printer->out('✔ ' . substr($file, strlen($rootDir) + 1) . "\n", 'info');
        }
    }

    /**
     * Removes everything until <!-- TEMPLATE --> from README.md
     * @param string $rootDir
     */
    private function removeReadmeSection(string $rootDir)
    {
        $readmeFile = $rootDir . '/README.md';
        $content = file_get_contents($readmeFile);
        $separator = "<!-- TEMPLATE -->\n\n";
        file_put_contents($readmeFile, substr($content, strpos($content, $separator) + strlen($separator)));
        $this->printer->out('✔ ' . "README.md template section removed\n", 'info');
    }

    private function success(array $values): void
    {
        $this->printer->success(
            implode(
                "\n",
                [
                    "┌──────────────────────────────────────────────────────────────┐",
                    "│ 🔽 All values have been replaced. Your next steps should be: │",
                    "└──────────────────────────────────────────────────────────────┘",
                ]
            )
        );
        $this->removeDevDirectory();
        $this->commitChanges();
        $this->connectServices($values);
        $this->startCoding($values);
        $this->installDependencies();
        $this->printer->success(
            implode(
                "\n",
                [
                    "┌──────────────────────────────────────────────────────────────┐",
                    "│ 🔼 Almost done! Please read your next steps above carefully! │",
                    "└──────────────────────────────────────────────────────────────┘",
                ]
            )
        );
    }

    private function removeDevDirectory()
    {
        $this->printer->out("🔸 Remove the dev directory:", 'info');
        $this->printer->display("\trm -rf ./dev\n");
    }

    private function commitChanges()
    {
        $this->printer->out("🔸 Commit and push initial version:", 'info');
        $this->printer->display("\tgit add -A && git commit -m 'Replace placeholders' && git push origin master\n");
    }

    private function connectServices(array $values)
    {
        $this->printer->out(
            implode(
                "\n",
                [
                    "🔸 Connect the following services with the Github repository:",
                    "   - Packagist to make it available with composer:",
                    "     https://packagist.org/packages/submit",
                    "   - Travis CI to run tests:",
                    "     https://travis-ci.org/organizations/{$values[':vendor']}/repositories",
                    "   - Scrutinizer for code quality and test coverage:",
                    "     https://scrutinizer-ci.com/g/new",
                    "   - (optional) Code Climate for more code quality metrics:",
                    "     https://codeclimate.com/github/repos/new",
                    "\n",
                ]
            ),
            'info'
        );
    }

    private function startCoding(array $values)
    {
        $this->printer->out("🔸 Install module IN A MAGENTO INSTALLATION to start developing:", 'info');
        $this->printer->display(
            "   composer require --prefer-source {$values[':vendor']}/{$values[':package']} dev-master\n"
        );
        $this->printer->out(
            "🔸 Now you have the Git repository in\n".
            "   vendor/{$values[':vendor']}/{$values[':package']}\n\n",
            'info'
        );
    }

    private function installDependencies()
    {
        $this->printer->out(
            "🔸 Install dev dependencies IN THAT DIRECTORY for automatic code quality checks:",
            'info'
        );
        $this->printer->display("   composer install");
    }
}


.travis/before_script.sh
#!/usr/bin/env bash

set -ex
trap '>&2 echo Error: Command \`$BASH_COMMAND\` on line $LINENO failed with exit code $?' ERR

# mock mail
sudo service postfix stop
echo # print a newline
smtp-sink -d "%d.%H.%M.%S" localhost:2500 1000 &
echo 'sendmail_path = "/usr/sbin/sendmail -t -i "' > ~/.phpenv/versions/$(phpenv version-name)/etc/conf.d/sendmail.ini

# adjust memory limit
echo 'memory_limit = -1' >> ~/.phpenv/versions/$(phpenv version-name)/etc/conf.d/travis.ini
phpenv rehash;

composer selfupdate

# clone main magento github repository
git clone --branch $MAGENTO_VERSION --depth=1 https://github.com/magento/magento2

# install Magento
cd magento2

# add composer package under test, composer require will trigger update/install
composer config minimum-stability dev
composer config repositories.travis_to_test git https://github.com/${TRAVIS_REPO_SLUG}.git
if [ ! -z $TRAVIS_TAG  ]
then
    composer require ${COMPOSER_PACKAGE_NAME}:${TRAVIS_TAG}
elif [ ! -z $TRAVIS_PULL_REQUEST_BRANCH ]
then
    # For pull requests, use the remote repository
    composer config repositories.travis_to_test git https://github.com/${TRAVIS_PULL_REQUEST_SLUG}.git
    composer require ${COMPOSER_PACKAGE_NAME}:dev-${TRAVIS_PULL_REQUEST_BRANCH}\#${TRAVIS_PULL_REQUEST_SHA}
else
    composer require ${COMPOSER_PACKAGE_NAME}:dev-${TRAVIS_BRANCH}\#${TRAVIS_COMMIT}
fi

# Add tests/src to autoload-dev on project level
php -r '$composer_json = json_decode(file_get_contents("composer.json"), true);
$composer_json["autoload-dev"]["psr-4"][":module-namespace\\:module-name\\"] = "vendor/:vendor/:package/src";
file_put_contents("composer.json", json_encode($composer_json));'
composer dumpautoload

# prepare for test suite
case $TEST_SUITE in
    integration)
        cp vendor/$COMPOSER_PACKAGE_NAME/tests/integration/phpunit.xml.dist dev/tests/integration/phpunit.xml

        cd dev/tests/integration

        # create database and move db config into place
        mysql -uroot -e '
            SET @@global.sql_mode = NO_ENGINE_SUBSTITUTION;
            CREATE DATABASE magento_integration_tests;
        '
        cp etc/install-config-mysql.php.dist etc/install-config-mysql.php
        # Remove AMQP configuration
        sed -i '/amqp/d' etc/install-config-mysql.php
        # Remove default root password
        sed -i 's/123123q//' etc/install-config-mysql.php

        cd ../../..

         echo "Wait for ElasticSearch on port 9200..."
         curl -X GET localhost:9200/_cluster/health?wait_for_status=yellow&timeout=10s

    ;;
    unit)
        cp vendor/$COMPOSER_PACKAGE_NAME/tests/unit/phpunit.xml.dist dev/tests/unit/phpunit.xml
    ;;
esac


dev/src/Config.php
<?php

namespace IntegerNet\ModuleTemplate;

use function array_merge as merge;

/**
 * These functions contain configuration for the init script
 */
class Config
{
    /**
     * @return string[] relative paths for files where placeholders will be replaced
     */
    public static function getFilesToUpdate(): array
    {
        return [
            'auth.json',
            'composer.json',
            '.travis.yml',
            '.travis/before_script.sh',
            'LICENSE',
            'README.md',
            'CONTRIBUTING.md',
            'registration.php',
            'src/ExampleClass.php',
            'src/etc/module.xml',
            'tests/integration/phpunit.xml.dist',
            'tests/integration/ExampleTest.php',
            'tests/unit/ExampleTest.php',
            'tests/unit/phpunit.xml.dist',
            'tests/src/ExampleTestDataBuilder.php',
        ];
    }

    /**
     * @return string[] Associative array with placeholders as keys and default values as values
     */
    public static function getDefaultVariables(): array
    {
        // Get author name and email from Git if possible
        $gitUserName = trim(shell_exec('git config --get user.name'));
        $gitUserEmail = trim(shell_exec('git config --get user.email'));

        // Get vendor and package from Github URI if possible (matches HTTPS and SSH)
        preg_match('#github\.com[:/]([^/]+)/(.*)\.git#', shell_exec('git remote get-url origin'), $matches);
        $vendor = $matches[1] ?? 'acme-inc';
        $package = $matches[2] ?? 'magento2-awesome-module';

        // Guess company, namespace and module name based on vendor and package
        $company = ucwords(str_replace('-', ' ', $vendor));
        $moduleNamespace = str_replace(' ', '', $company);
        $commonPrefixes = self::getCommonModulePrefixes($moduleNamespace);
        $moduleName = str_replace(merge([' '], $commonPrefixes), '', ucwords(str_replace(['-', '_'], ' ', $package)));

        return [
            ':magento-repo-public-key'  => '',
            ':magento-repo-private-key' => '',
            ':vendor'                   => $vendor,
            ':package'                  => $package,
            ':description'              => 'This module is awesome!',
            ':author-name'              => $gitUserName ?: 'John Doe',
            ':author-email'             => $gitUserEmail ?: 'john.doe@example.com',
            ':author-github'            => $vendor,
            ':module-namespace'         => $moduleNamespace,
            ':module-name'              => $moduleName,
            ':company'                  => $company,
            ':year'                     => (string)date('Y'),
        ];
    }

    /**
     * Exact PHP requirements are already handled by magento/framework itself, but making it explicit allows the module
     * author to tighten the constraints, e.g. make the module work with Magento 2.3 but still require at least PHP 7.3.
     *
     * Also, the supported PHP versions are used by {@see TravisJob::getConfiguration()} to generate Travis jobs.
     *
     * The following placeholder values will be generated automatically based on choosen compatible versions:
     *
     * @return string[][][] Associative array with Magento versions and dependencies for composer
     * @see Initialize::askMagentoCompatibility()
     *
     *  :php-constraint         Composer version constraint for PHP
     *  :framework-constraint   Composer version constraint for Magento framework
     *  :version-badge          Text for the "Magento" compatibility badge
     *
     */
    public static function getMagentoVersions(): array
    {
        return [
            '2.2' => [
                'php'               => ['~7.0', '~7.1', '~7.2'],
                'magento-framework' => ['^101.0.0'],
            ],
            '2.3' => [
                'php'               => ['~7.1', '~7.2', '~7.3'],
                'magento-framework' => ['^102.0.0'],
            ],
            '2.4' => [
                'php'               => ['~7.3', '~7.4'],
                'magento-framework' => ['^103.0.0'],
            ],
        ];
    }

    /**
     * package prefixes that should not be part of the module name (capitalized because used after ucwords())
     *
     * This covers e.g.:
     *
     *      magento2-
     *      magento2-module
     *
     * @param string $moduleNamespace
     * @return string[]
     */
    private static function getCommonModulePrefixes(string $moduleNamespace): array
    {
        return ['Magento2', 'Module', $moduleNamespace];
    }
}
